### PR TITLE
DS-2579: Accepting string labels

### PR DESCRIPTION
**Pull request by sugi-chan**
reason: allow octopod to take string labels as well as encoded labels. Basically the strategy is to use sklearn encoders to handle the incoming data in the dataset classes. The dataset classes then pass the label mappings to the dataloader class and when the learner is created it checks the train and val data loaders to make sure the label mappings are the same (a potential issue is that since the encoders are run on each dataset you might not have a certain class represented and so your train and val sets would have different label maps... which would be bad)

It looks long but is mostly tests... 4 for each of the image, text, and ensemble datasets. 2 for multi class and 2 for multi label. testing to make sure it loads with both string and ints and then checking to make sure it catches errors in loading if the label mappings are different between train and val datasets.

Note: I could just pass the class lists back and forth rather than the mappings I make. And I think they are ordered alphabetically by the encoder... But am passing the mapping bc I kinda like having it map to a node number so it could be used in the situation of a predict function where we wouldn't have to pass the sklearn label encoder around and save that places to use the inverse transform function in a preduct function.

Pull Request Checklist
 - [x] All tests in the `tests` folder pass with a local build
 - [x] Pull request includes a description of why we are doing this
 - [x] Init files import new capabilities to appropriate level of package (if applicable)
 - [x] CHANGELOG has been updated
 - [x] Version in `_version.py` has been updated
 - [x] README has been updated (if applicable)
 - [x] `requirements.txt` and `requirements-dev.txt` have been recompiled if requirements in `setup.py` or `requirements-dev.in` changed.

### CODE REVIEW

File: octopod/dataloader.py [hunk 1]
```diff
@@ -44,3 +45,10 @@ def __len__(self):
             num_batches += len(self.loader_dict[key])
 
         return num_batches
+
+    def _gen_label_mappings(self):
+
+        mapping_dict = {}
+        for key in self.loader_dict.keys():
+            mapping_dict[key] = self.loader_dict[key].dataset.label_mapping
```
Comment (nathancooperjones) [inline octopod/dataloader.py:53]: [Optional] This is pretty picky on my end (sorry), but it might read a bit nicer if we refactor this to: 

```python
for key, loader_dict in self.loader_dict.items():
    mapping_dict[key] = loader_dict.dataset.label_mapping
```

File: octopod/ensemble/dataset.py [hunk 2]
```diff
@@ -92,6 +98,13 @@ def __getitem__(self, index):
     def __len__(self):
         return len(self.text_inputs)
 
+    def _encode_labels(self):
+        """Encodes y labels using sklearn to create allow for string or numeric inputs"""
+        le = preprocessing.LabelEncoder()
```
Comment (nathancooperjones) [inline octopod/ensemble/dataset.py:103]: I didn't realize that, if given numeric inputs, `LabelEncoder` will make sure the mapping dictionary is numeric. Really neat! 

```python
>>> from sklearn import preprocessing
>>> le = preprocessing.LabelEncoder()
>>> y = [3, 2, 1, 0, 0, 8, 4, 5, 6, 7]
>>> le.fit(y)
LabelEncoder()
>>> dict(zip(le.transform(le.classes_), le.classes_))
{0: 0, 1: 1, 2: 2, 3: 3, 4: 4, 5: 5, 6: 6, 7: 7, 8: 8}
```

File: octopod/ensemble/dataset.py [hunk 3]
```diff
@@ -92,6 +98,13 @@ def __getitem__(self, index):
     def __len__(self):
         return len(self.text_inputs)
 
+    def _encode_labels(self):
+        """Encodes y labels using sklearn to create allow for string or numeric inputs"""
```
Comment (nathancooperjones) [inline octopod/ensemble/dataset.py:102]:
```suggestion
        """Encodes string or numeric y labels to integers using LabelEncoder"""
```

File: octopod/ensemble/dataset.py [hunk 4]
```diff
@@ -147,8 +161,19 @@ def __getitem__(self, index):
...
+        label = list(self.label_encoder.transform(label)[0])
+
+        y_output = torch.FloatTensor(label)
 
         return {'bert_text': bert_text,
                 'full_img': full_img,
                 'crop_img': cropped_img}, y_output
+
+    def _encode_labels(self):
+        """Encodes y labels using sklearn to create allow for string or numeric inputs"""
```
Comment (nathancooperjones) [inline octopod/ensemble/dataset.py:174]:
```suggestion
        """Encodes string or numeric y labels to integers using MultiLabelBinarizer"""
```

File: octopod/ensemble/dataset.py (hunk 3 above)
Comment (nathancooperjones) [reply octopod/ensemble/dataset.py:102]: You don't have to use this exact language, but I think it could be nice to make it clear exactly what is happening here.

File: octopod/ensemble/dataset.py [hunk 5]
```diff
@@ -83,7 +86,10 @@ def __getitem__(self, index):
         full_img = self.transform(full_img)
         cropped_img = self.crop_transform(cropped_img)
 
-        y_output = torch.from_numpy(np.array(self.y[index])).long()
+        label = self.y[index]
+        label = self.label_encoder.transform([label])[0]
```
Comment (nathancooperjones) [inline octopod/ensemble/dataset.py:90]:
```suggestion
        label = self.label_mapping[label]
```

Comment (nathancooperjones) [reply octopod/ensemble/dataset.py:90]: If we make this change, is there a reason to save `self.label_encoder` at all? Can we only use the mapping and still get by?

File: octopod/ensemble/dataset.py [hunk 6]
```diff
@@ -104,9 +117,10 @@ class OctopodEnsembleDatasetMultiLabel(OctopodEnsembleDataset):
     img_inputs: pandas Series
         the paths to images to be used
     y: list
-        a list of binary encoded categories with length equal to number of
+        a list of binary encoded categories or strings with length equal to number of
         classes in the multi-label task. For a 4 class multi-label task
-        a sample list would be [1,0,0,1]
+        a sample list would be [1,0,0,1], which will be encoded
+        using a sklearn label encoder
```
Comment (nathancooperjones) [inline octopod/ensemble/dataset.py:123]: Can we also add an example here with a sample list of strings?

File: octopod/ensemble/dataset.py [hunk 7]
```diff
@@ -104,9 +117,10 @@ class OctopodEnsembleDatasetMultiLabel(OctopodEnsembleDataset):
     img_inputs: pandas Series
         the paths to images to be used
     y: list
-        a list of binary encoded categories with length equal to number of
+        a list of binary encoded categories or strings with length equal to number of
```
Comment (nathancooperjones) [inline octopod/ensemble/dataset.py:120]: Is the expected `y` here a list of lists?

File: octopod/ensemble/dataset.py [hunk 8]
```diff
@@ -18,7 +19,8 @@ class OctopodEnsembleDataset(Dataset):
     img_inputs: pandas Series
         the paths to images to be used
     y: list
-        A list of lists of dummy-encoded categories
+        A list of lists of dummy-encoded categories or strings,
+        which will be encoded using a sklearn label encoder
```
Comment (nathancooperjones) [inline octopod/ensemble/dataset.py:23]: Wait... is this right? I thought this just needs a list, and `OctopodEnsembleDatasetMultiLabel` requires a list of lists?

File: octopod/learner.py [hunk 9]
```diff
@@ -433,6 +434,15 @@ def _check_for_all_tasks(self, input_dict, tasks, input_str):
             raise Exception(f'make sure all tasks are contained in the {input_str} dictionary '
                             f'missing tasks are {missing_tasks}')
 
+    def _check_all_labels_present(self):
+        # check that all categories are in both the train and val sets for each task
+        for task in self.task_dict:
+            if self.train_dataloader.label_mappings[task] != self.val_dataloader.label_mappings[task]: # noqa
+                raise Exception(f'Mapping mismatch in {task} task. Check that all categories are '
```
Comment (nathancooperjones) [inline octopod/learner.py:441]: Nice descriptive error message!

Comment (nathancooperjones) [inline octopod/learner.py:441]: [Optional] It might be clearer to raise a `ValueError` here instead of a general `Exception` - what do you think?

File: octopod/learner.py [hunk 10]
```diff
@@ -66,6 +66,7 @@ def __init__(self,
         self.tasks = [*task_dict]
         self.loss_function_dict = self._get_loss_functions(loss_function_dict)
         self.metric_function_dict = self._get_metric_functions(metric_function_dict)
+        self._check_all_labels_present()
```
Comment (nathancooperjones) [inline octopod/learner.py:69]: Can we add a small note to the docstring above this to say that we expect the labels of `train_dataloader` to exactly match that of `val_dataloader`?

File: octopod/text/dataset.py [hunk 11]
```diff
@@ -12,7 +13,8 @@ class OctopodTextDataset(Dataset):
     x: pandas Series
         the text to be used
     y: list
-        A list of lists of dummy-encoded categories
+        A list of lists of dummy-encoded or string categories
+        will be encoded using an sklearn label encoder
```
Comment (nathancooperjones) [inline octopod/text/dataset.py:17]: Same question here about list of lists - I think I might just be confused hahaha so my bad

File: octopod/text/dataset.py [hunk 12]
```diff
@@ -42,13 +45,23 @@ def __getitem__(self, index):
...
+        label = self.label_encoder.transform([label])[0]
+
+        y_output = torch.from_numpy(np.array(label)).long()
 
         return x_input, y_output
 
     def __len__(self):
         return len(self.x)
 
+    def _encode_labels(self):
```
Comment (nathancooperjones) [inline octopod/text/dataset.py:58]: [Optional] If you want to spend time doing something for literally no benefit, it could be nice to make base classes for `Dataset`s and `MultiLabelDataset`s that implement common methods like this so we only need to change a single place in the future. I understand if you don't want to do that, but whenever I see an opportunity for inheritance, I lose all control.

File: tests/fixtures/learner_fixtures.py [hunk 13]
```diff
@@ -89,3 +92,252 @@ def test_train_val_loaders():
...
 
     return dataset_loader
+
+
+@pytest.fixture()
+def mixed_labels():
+    df_string = pd.DataFrame({'image_loc': ['sample_data/tonks.jpeg', 'sample_data/tonks.jpeg'],
+                              'category': ['cat', 'dog']})
+    df_int = pd.DataFrame({'image_loc': ['sample_data/tonks.jpeg', 'sample_data/tonks.jpeg'],
+                           'category': [1, 2]})
```
Comment (nathancooperjones) [inline tests/fixtures/learner_fixtures.py:102]: We seem to use this in multiple fixtures, so it might be nice to just make this its own fixture that other fixtures can use, I.e.: 

```python
@pytest.fixture()
def df_string():
      return pd.DataFrame({'image_loc': ['sample_data/tonks.jpeg', 'sample_data/tonks.jpeg'],
                           'fake_text': ['sample_data/tonks.jpeg', 'sample_data/tonks.jpeg'],
                           'category': ['cat', 'dog']})

... 

@pytest.fixture()
def mixed_labels(df_string, df_int):
    ... 
```

File: tests/fixtures/learner_fixtures.py [hunk 14]
```diff
@@ -89,3 +92,252 @@ def test_train_val_loaders():
...
+
+    return dataset_loader, dataset_loader2
+
+
+@pytest.fixture()
+def multi_label_mixed_labels():
+    df_string = pd.DataFrame({'image_loc': ['sample_data/tonks.jpeg', 'sample_data/tonks.jpeg'],
+                              'category': [['cat', 'dog'], ['dog']]})
+    df_int = pd.DataFrame({'image_loc': ['sample_data/tonks.jpeg', 'sample_data/tonks.jpeg'],
+                           'category': [[1, 2], [2]]})
```
Comment (nathancooperjones) [inline tests/fixtures/learner_fixtures.py:134]: Same here!

File: tests/fixtures/learner_fixtures.py [hunk 15]
```diff
@@ -89,3 +92,252 @@ def test_train_val_loaders():
...
+
+    dataloaders_dict = {
+        'task1': DataLoader(task1_dataset, batch_size=1, shuffle=True, num_workers=2),
+        'task2': DataLoader(task2_datset, batch_size=1, shuffle=True, num_workers=2),
+    }
+
+    dataset_loader = MultiDatasetLoader(loader_dict=dataloaders_dict)
+    dataset_loader2 = MultiDatasetLoader(loader_dict=dataloaders_dict)
+
+    return dataset_loader, dataset_loader2
```
Comment (nathancooperjones) [inline tests/fixtures/learner_fixtures.py:126]: What do you think about pulling this out as a helper function that lives in `learner_fixtures.py` that every fixture can call with the task datasets to get back the data loader?

File: tests/fixtures/learner_fixtures.py [hunk 16]
```diff
@@ -89,3 +92,252 @@ def test_train_val_loaders():
...
+        y=df_string['category'],
+        transform='train',
+        crop_transform='train'
+    )
+
+    task2_datset = OctopodImageDataset(
+        x=df_int['image_loc'],
+        y=df_int['category'],
+        transform='train',
+        crop_transform='train'
```
Comment (nathancooperjones) [inline tests/fixtures/learner_fixtures.py:115]: Does it matter that these are also `train` and not `val`?

File: tests/test_learner.py [hunk 17]
```diff
@@ -140,3 +142,254 @@ def test_learner_with_multi_class_and_label_acc(test_acc_dict_all_tasks_diff_los
...
+
+    model = ResnetForMultiTaskClassification(
+        new_task_dict=task_dict)
+
+    with pytest.raises(Exception) as e:
+        MultiInputMultiTaskLearner(model,
+                                   train_dataset_loader,
+                                   val_dataset_loader,
+                                   task_dict)
+    assert ('Mapping mismatch in task1 task. Check that all categories' in str(e.value)) is True
```
Comment (nathancooperjones) [inline tests/test_learner.py:184]: Nice test!

**Review by nathancooperjones — CHANGES_REQUESTED** [review]
This PR is awesome! I left some suggestions, some of which are overly nitpick-y and you can just ignore. Nice work so far!

**Comment by nathancooperjones** [comment]
Also @sugi-chan can we make a ticket to update tutorial notebooks to use this new label encoder strategy?

File: octopod/ensemble/dataset.py (hunk 2 above)
Comment (sugi-chan) [reply octopod/ensemble/dataset.py:103]: yeah I thought that eas an interesting output... so yeah when we have numeric inputs we have really boring mappings lol

File: octopod/ensemble/dataset.py (hunk 5 above)
Comment (sugi-chan) [reply octopod/ensemble/dataset.py:90]: yeah we could technically use the encoder. it is slightly more annoying w/ the multi-label one, but I can also adjust those

File: octopod/ensemble/dataset.py (hunk 7 above)
Comment (sugi-chan) [reply octopod/ensemble/dataset.py:120]: technically lists... I guess it also works with pandas columns.... so it might just be iterables we can index into.

File: octopod/ensemble/dataset.py (hunk 8 above)
Comment (sugi-chan) [reply octopod/ensemble/dataset.py:23]: yeah thats probably a typo from some pt in the past

File: octopod/learner.py (hunk 9 above)
Comment (sugi-chan) [reply octopod/learner.py:441]: fair

File: octopod/ensemble/dataset.py (hunk 8 above)
Comment (sugi-chan) [reply octopod/ensemble/dataset.py:23]: yeah looking at im guessing i swapped the y descriptions between the multi-label and regular datasets at some pt

File: octopod/text/dataset.py (hunk 12 above)
Comment (sugi-chan) [reply octopod/text/dataset.py:58]: decent future ticket... but im also less motivated by refactoring LOL

File: tests/fixtures/learner_fixtures.py (hunk 16 above)
Comment (sugi-chan) [reply tests/fixtures/learner_fixtures.py:115]: not really but can swap out

File: octopod/ensemble/dataset.py (hunk 5 above)
Comment (sugi-chan) [reply octopod/ensemble/dataset.py:90]: So I think for now we should leave it as using the tranforms from the sklearn encoders. For the multi-class ones it doesnt really matter. the multi-label biz I would probably just copy over the inverse_transform code into octopod 

https://github.com/scikit-learn/scikit-learn/blob/95119c13a/sklearn/preprocessing/_label.py#L660

File: octopod/learner.py [hunk 18]
```diff
@@ -51,6 +51,11 @@ class MultiTaskLearner(object):
         for that task.
 
         See `learner_utils.loss_utils_config` for examples.
+
+    NOTES
+    -----
+        The label mappings from the train and validation datasets for each task are expected to have
+        all the same labels present maps. If they do not then an error will be thrown.
```
Comment (nathancooperjones) [inline octopod/learner.py:58]:
```suggestion
        all the same labels present maps. If they do not then a ValueError will be thrown.
```

File: octopod/ensemble/dataset.py (hunk 7 above)
Comment (nathancooperjones) [reply octopod/ensemble/dataset.py:120]: Should this be a `list of lists` too to match the input for the text and vision datasets?

File: octopod/text/dataset.py [hunk 19]
```diff
@@ -72,10 +72,12 @@ class OctopodTextDatasetMultiLabel(OctopodTextDataset):
...
     y: list
-        a list of binary encoded categories or string categories
+        a list of lists of binary encoded categories or string categories
         with length equal to number of classes in the multi-label task.
-        For a 4 class multi-label task a sample list would be [1,0,0,1]
-        will be encoded using an sklearn label encoder
+        For a 4 class multi-label task a sample list would be [1,0,0,1],
+        A string example would be ['cat','dog'],
+        (if the classes were ['cat','frog','rabbit','dog]), which will be encoded
+        using a sklearn label encoder to [1,0,0,1].
```
Comment (nathancooperjones) [inline octopod/text/dataset.py:80]: Nice example!!

File: tests/fixtures/learner_fixtures.py [hunk 20]
```diff
@@ -108,30 +132,20 @@ def mixed_labels():
...
-    dataloaders_dict = {
-        'task1': DataLoader(task1_dataset, batch_size=1, shuffle=True, num_workers=2),
-        'task2': DataLoader(task2_datset, batch_size=1, shuffle=True, num_workers=2),
-    }
-
-    dataset_loader = MultiDatasetLoader(loader_dict=dataloaders_dict)
-    dataset_loader2 = MultiDatasetLoader(loader_dict=dataloaders_dict)
-
-    return dataset_loader, dataset_loader2
+    return gen_train_val_dataloaders_helper(task1_dataset, task2_dataset)
```
Comment (nathancooperjones) [inline tests/fixtures/learner_fixtures.py:142]: Nice!!!!!

File: octopod/ensemble/dataset.py (hunk 5 above)
Comment (nathancooperjones) [reply octopod/ensemble/dataset.py:90]: Makes sense to me!

File: octopod/learner.py [hunk 21]
```diff
@@ -51,6 +51,11 @@ class MultiTaskLearner(object):
         for that task.
 
         See `learner_utils.loss_utils_config` for examples.
+
+    NOTES
```
Comment (nathancooperjones) [inline octopod/learner.py:55]: [Optional] I think it's also okay to put this explanation in the docstring under `val_dataloader ` above this unless you like it better in `NOTES`.

File: octopod/learner.py (hunk 18 above)
Comment (sugi-chan) [reply octopod/learner.py:58]: ugh forgetting to update notes

File: octopod/learner.py (hunk 21 above)
Comment (sugi-chan) [reply octopod/learner.py]: I just dumped notes into both the train and val dataloader notes LOL

**Review by nathancooperjones — APPROVED** [review]
Woo-hoo!!!! LGTM! 🎉
